### ISSUE TITLE
Duplicate files/directories starting from 0.22.0

### ISSUE THREAD

**Issue by crsndr** [kahing/goofys#487]
Dear goofys developers,
we're using an internal developed implementation of S3. We started to use toward September last month version 0.20.0 and after updating to 0.22.0 and more recently to 0.23.1 we've noticed that in some cases some directories seems to be shown twice. For example:

`2020/01/14 16:32:42.719500 s3.DEBUG DEBUG: Request s3/HeadObject Details:
---[ REQUEST POST-SIGN ]-----------------------------
HEAD /imaging-public/world_availability HTTP/1.1
Host: hh.fire.sdo.ebi.ac.uk
User-Agent: aws-sdk-go/1.17.13 (go1.13.1; linux; amd64)

-----------------------------------------------------
2020/01/14 16:32:42.719550 s3.DEBUG DEBUG: Request s3/ListObjectsV2 Details:
---[ REQUEST POST-SIGN ]-----------------------------
GET /imaging-public?delimiter=%2F&list-type=2&max-keys=1&prefix=world_availability%2F HTTP/1.1
Host: hh.fire.sdo.ebi.ac.uk
User-Agent: aws-sdk-go/1.17.13 (go1.13.1; linux; amd64)
Accept-Encoding: identity

-----------------------------------------------------
2020/01/14 16:32:42.719512 s3.DEBUG DEBUG: Request s3/HeadObject Details:
---[ REQUEST POST-SIGN ]-----------------------------
HEAD /imaging-public/world_availability/ HTTP/1.1
Host: hh.fire.sdo.ebi.ac.uk
User-Agent: aws-sdk-go/1.17.13 (go1.13.1; linux; amd64)

-----------------------------------------------------
2020/01/14 16:32:42.844194 s3.DEBUG DEBUG: Response s3/HeadObject Details:
---[ RESPONSE ]--------------------------------------
HTTP/1.1 404 Not Found
Connection: close
Cache-Control: no-cache, no-store, max-age=0, must-revalidate
Content-Type: application/json;charset=UTF-8
Date: Tue, 14 Jan 2020 16:32:42 GMT
Expires: 0
Pragma: no-cache
Set-Cookie: JSESSIONID=D93DA3379C5C8E4CD28B7931D9A60328; Path=/; HttpOnly
X-Content-Type-Options: nosniff
X-Frame-Options: DENY
X-Xss-Protection: 1; mode=block

-----------------------------------------------------
2020/01/14 16:32:42.844227 s3.DEBUG DEBUG: Validate Response s3/HeadObject failed, not retrying, error NotFound: Not Found
        status code: 404, request id: , host id: 
2020/01/14 16:32:42.844287 s3.DEBUG HEAD world_availability = no such file or directory
2020/01/14 16:32:42.846346 s3.DEBUG DEBUG: Response s3/HeadObject Details:
---[ RESPONSE ]--------------------------------------
HTTP/1.1 404 Not Found
Cache-Control: no-cache, no-store, max-age=0, must-revalidate
Date: Tue, 14 Jan 2020 16:32:42 GMT
Expires: 0
Pragma: no-cache
Set-Cookie: JSESSIONID=D2504B29659548E927AEBE2C83FF32F9; Path=/; HttpOnly
X-Content-Type-Options: nosniff
X-Frame-Options: DENY
X-Xss-Protection: 1; mode=block
Content-Length: 0

-----------------------------------------------------
2020/01/14 16:32:42.846400 s3.DEBUG DEBUG: Validate Response s3/HeadObject failed, not retrying, error NotFound: Not Found
        status code: 404, request id: , host id: 
2020/01/14 16:32:42.846427 s3.DEBUG HEAD world_availability/ = no such file or directory
2020/01/14 16:32:42.860310 s3.DEBUG DEBUG: Response s3/ListObjectsV2 Details:
---[ RESPONSE ]--------------------------------------
HTTP/1.1 200 OK
Content-Length: 341
Cache-Control: no-cache, no-store, max-age=0, must-revalidate
Content-Type: text/plain;charset=UTF-8
Date: Tue, 14 Jan 2020 16:32:42 GMT
Expires: 0
Pragma: no-cache
Server: Fire-S3-Gateway
Set-Cookie: JSESSIONID=0E02CBB49D33E77DF242E95C869F002D; Path=/; HttpOnly
X-Amz-Bucket-Region: us-east-1
X-Content-Type-Options: nosniff
X-Frame-Options: DENY
X-Xss-Protection: 1; mode=block

-----------------------------------------------------
2020/01/14 16:32:42.860351 s3.DEBUG &{[world_availability/10002/] [] 0xc00067cf28 true : }
2020/01/14 16:32:42.860724 s3.DEBUG DEBUG: Request s3/ListObjectsV2 Details:
---[ REQUEST POST-SIGN ]-----------------------------
GET /imaging-public?delimiter=%2F&list-type=2&max-keys=1&prefix=world_availability%2F10304%2F HTTP/1.1
Host: hh.fire.sdo.ebi.ac.uk
User-Agent: aws-sdk-go/1.17.13 (go1.13.1; linux; amd64)
Accept-Encoding: identity

-----------------------------------------------------
2020/01/14 16:32:42.860770 s3.DEBUG DEBUG: Request s3/HeadObject Details:
---[ REQUEST POST-SIGN ]-----------------------------
HEAD /imaging-public/world_availability/10304/ HTTP/1.1
Host: hh.fire.sdo.ebi.ac.uk
User-Agent: aws-sdk-go/1.17.13 (go1.13.1; linux; amd64)

-----------------------------------------------------
2020/01/14 16:32:42.860754 s3.DEBUG DEBUG: Request s3/HeadObject Details:
---[ REQUEST POST-SIGN ]-----------------------------
HEAD /imaging-public/world_availability/10304 HTTP/1.1
Host: hh.fire.sdo.ebi.ac.uk
User-Agent: aws-sdk-go/1.17.13 (go1.13.1; linux; amd64)

-----------------------------------------------------
2020/01/14 16:32:42.870455 s3.DEBUG DEBUG: Response s3/HeadObject Details:
---[ RESPONSE ]--------------------------------------
HTTP/1.1 404 Not Found
Connection: close
Cache-Control: no-cache, no-store, max-age=0, must-revalidate
Content-Type: application/json;charset=UTF-8
Date: Tue, 14 Jan 2020 16:32:42 GMT
Expires: 0
Pragma: no-cache
Set-Cookie: JSESSIONID=294285525664F3EC4FE8073541CEDAC1; Path=/; HttpOnly
X-Content-Type-Options: nosniff
X-Frame-Options: DENY
X-Xss-Protection: 1; mode=block

-----------------------------------------------------
2020/01/14 16:32:42.870482 s3.DEBUG DEBUG: Validate Response s3/HeadObject failed, not retrying, error NotFound: Not Found
        status code: 404, request id: , host id: 
2020/01/14 16:32:42.870505 s3.DEBUG HEAD world_availability/10304 = no such file or directory
2020/01/14 16:32:42.873089 s3.DEBUG DEBUG: Response s3/ListObjectsV2 Details:
---[ RESPONSE ]--------------------------------------
HTTP/1.1 200 OK
Content-Length: 352
Cache-Control: no-cache, no-store, max-age=0, must-revalidate
Content-Type: text/plain;charset=UTF-8
Date: Tue, 14 Jan 2020 16:32:42 GMT
Expires: 0
Pragma: no-cache
Server: Fire-S3-Gateway
Set-Cookie: JSESSIONID=4E950958F477F1ADCEF94122B08A0F13; Path=/; HttpOnly
X-Amz-Bucket-Region: us-east-1
X-Content-Type-Options: nosniff
X-Frame-Options: DENY
X-Xss-Protection: 1; mode=block

-----------------------------------------------------
2020/01/14 16:32:42.873164 s3.DEBUG &{[world_availability/10304/data/] [] 0xc0003fffa8 true : }
2020/01/14 16:32:42.875923 s3.DEBUG DEBUG: Request s3/ListObjectsV2 Details:
---[ REQUEST POST-SIGN ]-----------------------------
GET /imaging-public?delimiter=%2F&list-type=2&prefix=world_availability%2F10304%2F HTTP/1.1
Host: hh.fire.sdo.ebi.ac.uk
User-Agent: aws-sdk-go/1.17.13 (go1.13.1; linux; amd64)
Accept-Encoding: identity

-----------------------------------------------------
2020/01/14 16:32:42.885312 s3.DEBUG DEBUG: Response s3/ListObjectsV2 Details:
---[ RESPONSE ]--------------------------------------
HTTP/1.1 200 OK
Content-Length: 352
Cache-Control: no-cache, no-store, max-age=0, must-revalidate
Content-Type: text/plain;charset=UTF-8
Date: Tue, 14 Jan 2020 16:32:42 GMT
Expires: 0
Pragma: no-cache
Server: Fire-S3-Gateway
Set-Cookie: JSESSIONID=4E7E7A41D5A1C5409936FD9353CC5D97; Path=/; HttpOnly
X-Amz-Bucket-Region: us-east-1
X-Content-Type-Options: nosniff
X-Frame-Options: DENY
X-Xss-Protection: 1; mode=block

-----------------------------------------------------
2020/01/14 16:32:42.885380 s3.DEBUG &{[world_availability/10304/data/] [] 0xc00021aa48 true : }
2020/01/14 16:32:42.885633 s3.DEBUG DEBUG: Request s3/ListObjectsV2 Details:
---[ REQUEST POST-SIGN ]-----------------------------
GET /imaging-public?continuation-token=fYnJevcN9z-OBJECT%230&delimiter=%2F&list-type=2&prefix=world_availability%2F10304%2F HTTP/1.1
Host: hh.fire.sdo.ebi.ac.uk
User-Agent: aws-sdk-go/1.17.13 (go1.13.1; linux; amd64)
Accept-Encoding: identity

-----------------------------------------------------
2020/01/14 16:32:42.900108 s3.DEBUG DEBUG: Response s3/ListObjectsV2 Details:
---[ RESPONSE ]--------------------------------------
HTTP/1.1 200 OK
Content-Length: 487
Cache-Control: no-cache, no-store, max-age=0, must-revalidate
Content-Type: text/plain;charset=UTF-8
Date: Tue, 14 Jan 2020 16:32:42 GMT
Expires: 0
Pragma: no-cache
Server: Fire-S3-Gateway
Set-Cookie: JSESSIONID=2FD06E70A44C7C90AA493352D9C1E622; Path=/; HttpOnly
X-Amz-Bucket-Region: us-east-1
X-Content-Type-Options: nosniff
X-Frame-Options: DENY
X-Xss-Protection: 1; mode=block

-----------------------------------------------------
2020/01/14 16:32:42.900146 s3.DEBUG &{[] [world_availability/10304/10304.xml: 4235] 0xc0003fe628 false : }
2020/01/14 16:32:43.949442 s3.DEBUG DEBUG: Response s3/HeadObject Details:
---[ RESPONSE ]--------------------------------------
HTTP/1.1 404 Not Found
Cache-Control: no-cache, no-store, max-age=0, must-revalidate
Date: Tue, 14 Jan 2020 16:32:43 GMT
Expires: 0
Pragma: no-cache
Set-Cookie: JSESSIONID=88636794EC5AE0CD6499D70891D24099; Path=/; HttpOnly
X-Content-Type-Options: nosniff
X-Frame-Options: DENY
X-Xss-Protection: 1; mode=block
Content-Length: 0

-----------------------------------------------------
2020/01/14 16:32:43.949469 s3.DEBUG DEBUG: Validate Response s3/HeadObject failed, not retrying, error NotFound: Not Found
        status code: 404, request id: , host id:`

the correspondent command on the command line is:

`ls -l imaging-test/world_availability/10304/
total 8
drwxr-xr-x. 2 root root 4096 Jan 14 16:30 data
drwxr-xr-x. 2 root root 4096 Jan 14 16:30 data`

Oddly if the result is still on the cache. and if the cache hasn't expired, a second execution of the command produce the expected result:

`ls -l imaging-test/world_availability/10304/
total 9
-rw-r--r--. 1 root root 4235 Oct  4 11:31 10304.xml
drwxr-xr-x. 2 root root 4096 Jan 14 16:30 data`

which doesn't seem to produce any output on the debug.
If you need any more detail or test please let me know.

**Comment by kahing** [comment]
Is the debug log truncated in anyway? `world_availability/10304/` only contains `data/` and `10304.xml`?

**Comment by crsndr** [comment]
Yes it only contains those two elements: data/ and 10304.xml. I checked now and I see the same behaviour there:

ls -l /xfer/ro/empiar/fire/pub/world_availability/10304/
total 8
drwxr-xr-x. 2 root root 4096 Jan 15 15:55 data
drwxr-xr-x. 2 root root 4096 Jan 15 15:55 data

and at a second execution with the result still in cached:

ls -l /xfer/ro/empiar/fire/pub/world_availability/10304/
total 9
-rw-r--r--. 1 root root 4235 Oct  4 11:31 10304.xml
drwxr-xr-x. 2 root root 4096 Jan 15 15:55 data

If I wait for cache expiration (this is the setting on the mount--stat-cache-ttl=60s,--type-cache-ttl=60s) I get again the two data/ in the listing.

**Comment by kahing** [comment]
does your s3 implementation return objects in alphabetical order in the same way as aws s3?

**Comment by stuio** [comment]
Hi,

Thanks for the response. The implementation leverages on a legacy data model which makes returning Prefixes/Objects efficiently (i.e. using pagination) challenging. In a perhaps misguided attempt at mitigation we return Prefixes in one request, and Objects in a second request. So for paths with one or more Prefixes we require a second request to retrieve the Objects. Goofys is understanding our 'truncated' response correctly and initiating a second request with the provided continuation token to retrieve the Objects. 

Within each request the responses are ordered alphanumerically, but given we return Prefixes and Objects in separate requests there is no guarantee that the listing is alphanumeric over the two different types. Might this result in the duplication artefact we are seeing?

Thanks

**Comment by kahing** [comment]
Yes that would definitely cause the issue. I have some in-progress work to remove that requirement but currently no ETA

**Comment by stuio** [comment]
OK, that's great info, thanks. We'll have a look again at our implementation and see if we can guarantee ordering.
